build.gradle

```diff
@@ -189,6 +189,9 @@ project('segmentstore:storage') {
 project('segmentstore:storage:impl') {
     dependencies {
         compile group: 'org.apache.bookkeeper', name: 'bookkeeper-server', version: bookKeeperVersion, withoutLog4j
+        // https://mvnrepository.com/artifact/org.apache.curator/curator-x-async
+        compile group: 'org.apache.curator', name: 'curator-x-async', version: '4.0.0'
```

Comment:
So, we already use the CuratorFramework in this project. Is there really a pressing need to add another dependency on another version of this?

Reply:
After looking at the PR, it seems that this library doesn't add much value over what we already have. My recommendation is to drop this (as we already have a lot of dependencies already) and use an Executor to make things async. Once #1980 is done you won't have to worry about this at all, since Tier2 is essentially "sync" and it's made async by means of an executor.

Reply:
`curator-x-async` is not another version of curator. It is just an async wrapper on top of the existing curator.

Reply:
The option was to implement hand rolled executor based async interface or use an existing library. My evaluation was that it will be less time consuming and bug-free to reuse an already existing code. I think `curator-x-async` is a good addition to the dependencies.

Reply:
Our list of dependencies remains insane, if we can drop this one, then definitely better. I had a look at where we are using it in this PR and it isn't clear to me it is strictly necessarily or even that it requires much work to not use it. See my comment below, please.

Reply:
I took a closer look at the Async CuratorFramework. It is simply a wrapper on top of the sync CuratorFramework and it uses a Thread Pool to execute all sync operations. This is no different than what we'd be doing ourselves, so from that respect it doesn't provide any sort of value.

On top of that, in PR #1980 I have refactored all Storage implementations to actually implement a new interface called `SyncStorage`, which is the same as `Storage`, but all synchronous. That was because all of our implementations were actually synchronous and each was doing its own conversion to a CompletableFuture. With that PR, your BK Tier2 impl would look simpler and all you need to do in the BK Storage Factory is to wrap it in an `AsyncStorageWrapper` which makes it work on a thread pool.

Reply:
`With that PR, your BK Tier2 impl would look simpler and all you need to do in the BK Storage Factory is to wrap it in an AsyncStorageWrapper which makes it work on a thread pool.`

If we are going to  replace this implementation, does it make sense to spend time rolling out our own executor based solution? I'd rather reuse the existing library and then replace it with sync version later.

Reply:
I understand your concern, but between having something of the sort `CompletableFuture.runAsync/supplyAsync` and pulling yet another dependency that does pretty much the same, I prefer the former.